crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+			ensure!(current_owner == origin, Error::<T>::NoPermission);
+
+			ensure!(!<Listings<T>>::contains_key(&collection_id, token_id), Error::<T>::TokenListingProtection);
+
+			let listing_end_block = <frame_system::Module<T>>::block_number().saturating_add(T::DefaultListingDuration::get());
+			ListingEndSchedule::<T>::mutate(listing_end_block, |schedule| schedule.push((collection_id.clone(), token_id)));
+			let listing = Listing::<T>::DirectSale(
+				DirectSaleListing::<T> {
+					payment_asset,
+					fixed_price,
+					close: listing_end_block,
+					buyer: buyer.clone(),
+				}
+			);
+			Listings::insert(&collection_id, token_id, listing);
+			Self::deposit_event(RawEvent::DirectSaleListed(collection_id, token_id, buyer, payment_asset, fixed_price));
+		}
+
+		/// Buy an NFT for its listed price, must be listed for sale
+		#[weight = T::WeightInfo::direct_purchase()]
+		#[transactional]
+		fn direct_purchase(origin, collection_id: CollectionId, token_id: T::TokenId) {
+			let origin = ensure_signed(origin)?;
+			ensure!(<Listings<T>>::contains_key(&collection_id, token_id), Error::<T>::NotForDirectSale);
+
+			if let Some(Listing::DirectSale(listing)) = Self::listings(&collection_id, token_id) {
+
+				match listing.buyer {
+					// if buyer is specified in the listing, then `origin` must be buyer
+					Some(buyer) => ensure!(origin == buyer, Error::<T>::NoPermission),
+					None => (),
+				};
+
+				let current_owner = Self::token_owner(&collection_id, token_id);
+
+				// if there are no custom royalties, fallback to default if it exists
+				let royalties_schedule = if let Some(royalties_schedule) = Self::token_royalties(&collection_id, token_id) {
+					royalties_schedule
+				} else {
+					Self::collection_royalties(&collection_id).unwrap_or_else(Default::default)
+				};
+
+				let royalty_fees = royalties_schedule.calculate_total_entitlement();
+				if royalty_fees.is_zero() {
+					// full proceeds to seller/`current_owner`
+					T::MultiCurrency::transfer(&origin, &current_owner, Some(listing.payment_asset), listing.fixed_price, ExistenceRequirement::AllowDeath)?;
+				} else {
+					// withdraw funds from buyer, split between royalty payments and seller
+					let for_royalties = royalty_fees * listing.fixed_price;
+					let for_seller = listing.fixed_price - for_royalties;
+
+					let mut imbalance = T::MultiCurrency::withdraw(&origin, Some(listing.payment_asset), listing.fixed_price, WithdrawReason::Transfer.into(), ExistenceRequirement::AllowDeath)?;
+					imbalance = imbalance.offset(T::MultiCurrency::deposit_into_existing(&current_owner, Some(listing.payment_asset), for_seller)?).map_err(|_| Error::<T>::InternalPayment)?;
+					for (who, entitlement) in royalties_schedule.entitlements.into_iter() {
+						let amount = entitlement * listing.fixed_price;
+						imbalance = imbalance.offset(T::MultiCurrency::deposit_into_existing(&who, Some(listing.payment_asset), amount)?).map_err(|_| Error::<T>::InternalPayment)?;
```

Comment:
I would suggest paying the royalties first while skipping the payments to the accounts who don't exist and then the rest/remaining amount to the seller.

Reply:
I am particularly concerned with the case where Round(ax+ay) != Round(ax) + Round(ay), where a is the price and x and y are the royalty shares. In the case that we calculate the royalties first and pay the rest to the seller, there wouldn't be any different calculation for the share of the seller and thus we wouldn't encounter this issue.

Reply:
should be fixed now